### ISSUE TITLE
long-form printing of (mutable) string dict doesn't make sense

### ISSUE THREAD

**Issue by shriram** [brownplt/pyret-lang#965]
When you click to alter the view, a string dict prints as:

![image](https://cloud.githubusercontent.com/assets/75030/21896652/22480c0e-d8b5-11e6-9206-0a7cdf4eea68.png)

This doesn't really make sense as output. There are 5 paired items, not 10 individual ones. 

[Not the highest priority, but it's a bit confusing and unsavory.]

**Comment by blerner** [comment]
This is an immediate consequence of the current "lousy" constructor syntax for dictionaries. The idea was, since we write constructor notation as `[string-dict: "a", 2, "b", 3]`, then the rendering of that matches all other constructor syntaxes, for consistency. For most collections, having numbered items makes sense...

Once we have a tuples-based constructor syntax for dictionaries, we can revisit this rendering and have it make more sense.

**Comment by blerner** [comment]
@shriram I can hack in a fix for this (i.e. a special-purpose renderer for string-dicts, much like I did for colors).  Is this annoying enough that you'd want it fixed by this summer?

**Comment by shriram** [comment]
@blerner No, no need to hack something just for this summer. Thanks.

**Comment by shriram** [comment]
(But I assume the PR will remain open until there's a better solution.)

**Comment by jpolitz** [comment]
FWIW, we have tuple-based dictionaries in `str-dict` ( @sarahkf implemented it last summer).  They should implement the same interface as `string-dict`.

```
include str-dict

msd = [mutable-string-dict: {"a"; 5}]
isd = [string-dict: {"a"; 10}]
```

It's mostly a culture, support code, and documentation/deprecation issue to make the shift happen.

**Comment by blerner** [comment]
Good point, we should try to migrate people over to that library.  On the other hand, looking at those... those tuples are *ugly* with a lot of line noise, and the constant alternation between semicolons (between key and value) and commas (between key-value pairs) is finicky to remember.  I don't think I'd *want* to print those out long-form as
```
[string-dict:
Item 1: {"a"; 5},
Item 2: {"b"; 6},
...
]
```

A long time ago, I'd floated the idea that `a => b` could be syntactic sugar for `{a; b}`.  (I'm borrowing the idea from Perl/Python/Ruby-like syntax.)  We could even limit it to inside of `[constr: ...]` notation, and that'd avoid any parsing issues I can think of. (I'm not claiming it's pedagogically easy; that's a separate concern, but tuple syntax is another dependency to teach, too.)  Using this syntax, I'd write
```
[string-dict: "a" => 5, "b" => 6] is
[string-dict:
Item 1: "a" => 5,
Item 2: "b" => 6
]
```
and subjectively that looks far less noisy to me.

**Comment by shriram** [comment]
Is there a general form of this where any [<foo>: …] constructor can use it
in some well-defined way?

[I'm thinking, "how can we define the automaton macro?" <-;]

**Comment by blerner** [comment]
Literally, `a => b` will by pure sugar, and parse as a `s-tuple` (and I might extend the s-tuple AST variant to include a flag saying "pretty-print yourself using pointy-style", so that re-pretty-printing a program will preserve the style).  So any `[constr: ...]` that processes a bunch of pairs will work just fine with it, and it'll be up to the data definition that `constr` defines to implement an `_output` method that uses pointy-style pairs.  (That'd be a new value-skeleton constructor, basically.)

**Comment by shriram** [comment]
And it'll associate (right), so I can write

 [foo:
    a => b => c,
    d => e => f]

?

**Comment by blerner** [comment]
What do you want that to mean?  `[foo: {a; {b; c}}, {d; {e; f}}]`?  

My initial instinct is to say, no, it doesn't actually nest at all.  It'd be a special form that constructor-syntax has access to, if desired.  So it'd be specific to key-value--like constructions.  We might be able to make it more flexible than that, but then you start blurring the lines of what the => token means...

**Comment by shriram** [comment]
I'm fine with that too. I just want us to be clear on what you're proposing.
